### ISSUE TITLE
Moving the refman infrastructure to Alectryon?

### ISSUE THREAD

**Issue by Zimmi48** [coq/coq#16779]
Now that SerAPI is in the Coq CI and that the refman is split into a separate package into the Coq repository, we could seriously consider whether we would like to make the refman depend on Alectryon instead of maintaining our custom coqrst infrastructure.

IMHO, it would not be an issue for the Coq refman to depend on Alectryon and SerAPI, because most users do not need to build the refman themselves, and SerAPI is anyway included in the Coq Platform (should Alectryon be included there as well BTW?).

cc @coq/build-maintainers @coq/ci-maintainers @coq/doc-maintainers (@cpitclaudel in particular) @palmskog

I do not have a clear view of what amount of efforts would be required for such a switch. Would it be easy to port the missing features from coqrst to Alectryon? Or would we actually keep a small coqrst part with the Sphinx extensions that we use and that are not part of Alectryon (like `prodn`, `cmd`, `tacn`)?

**Comment by jfehrle** [comment]
How do users benefit from switching to Alectryon?

I think there are many other Coq improvements that would benefit users more.  Refactoring/replacing existing code is not an end in itself; it makes sense when the current code blocks bug fixes or new functionality.  Refactoring/replacing may break existing functionality, which may not be detected until much later.  What are the coqrst issues that this change would address?

From the description so far, I don't want this to divert any of my time.

**Comment by Zimmi48** [comment]
The main point for such a switch is that since Alectryon exists and will be the thing that users use, it becomes clear that coqrst will never be used for anything other than the Coq refman. Therefore, assuming we do not switch, we have the choice of having a legacy infrastructure that will never evolve (because no one wants to invest time in it) or making this infrastructure evolve at the price of duplicated work with Alectryon. In particular, Alectryon has already a few nice bells and whistles for displaying proof steps in Coq examples and could also be used to get a unified documentation for both the refman and the stdlib, but I don't see this happening if it means having to reimplement these features on top of coqrst. Alectryon is bound to evolve more and get better over time, because of its large user base (it goes even beyond Coq since it also supports Lean and other backends, yet many features are implemented only once and immediately usable on the various backends).

If we use a standard tool for the Coq documentation, it also means that it will be easier to encourage other Coq Platform projects to use the same, and eventually provide a unified documentation of the Coq Platform (even if it comes from various repositories that are maintained independently).

But at this stage, I'm just looking for an evaluation of the cost of switching (and I hope that @cpitclaudel could give some cost evaluation off the top of his head). If it is small, it may make sense to switch soon. If it is larger, then it may make sense to postpone such an endeavor.

**Comment by jfehrle** [comment]
It makes sense to look at it if it's a question of investing in coqrst enhancements or using/adapting Alectryon, .  It would be good to first come up with a wish list for improved documentation infrastructure (requirements and benefits) and then consider how well Alectryon can address the list, identify things we'd need to implement, any problem areas, etc.  Switching to Alectryon seems like it would not be easily to reverse, hence good to plan it out.

Does Alectryon provide better formatting and perhaps dynamic HTML effects, eg tooltips on proofs or the ability to expand/hide some material?  If we use Alectryon, the input is still an rst file (with minimal conversion of input) and the output is HTML?  Will it work with the pdf form of the manual?  Does it support the same or better indexing capabilities?  It would be interesting to understand how easy it is to extend various aspects of Alectryon.

I have a few thoughts about how to present stdlib documentation.  We should also think about documentation for external packages and how to do searching and HTML linking across multiple packages.

**Comment by cpitclaudel** [comment]
Hi folks! Addressing questions by @jfehrle first, but even before that taking a step back.  Jim, Alectryon isn't a replacement for Sphinx, it's a plugin for it.   With that it mind:

> Does Alectryon provide better formatting and perhaps dynamic HTML effects, eg tooltips on proofs or the ability to expand/hide some material? 

Yes, see the website for examples

> If we use Alectryon, the input is still an rst file (with minimal conversion of input) 

Yes, but I would expect no conversion of input to start with.  Alectryon can be used as a library and reuse the existing annotations.

> and the output is HTML? 

Yes, or LaTeX; the output for everything that's not the code is generated by Sphinx.

> Will it work with the pdf form of the manual? 

Yes; people have written full books with it at this point.

> Does it support the same or better indexing capabilities? 

It supports no indexing, because all the indexing is done by Sphinx.  That would not change.

> It would be interesting to understand how easy it is to extend various aspects of Alectryon.

I can only point to the website and the paper for that question :)

> Switching to Alectryon seems like it would not be easily to reverse, hence good to plan it out.

The Alectryon distribution contains a `minimal.py` files that's <100 lines that implements a shim of the Sphinx plugin (which does no code formatting at all, just passes things through), so in the worst case reverting should be pretty easy.

But to reiterate, Alectryon is essentially just a fancier reimplementation of coqrst.

Now on to @Zimmi48's main question:

> But at this stage, I'm just looking for an evaluation of the cost of switching (and I hope that @cpitclaudel could give some cost evaluation off the top of his head). 

I'd say about one day of work to replace the coqrst API, potentially one day if there's any missing feature (the main one I think would be hidden commands like `reset`), and one day to package everything properly?  Quite reasonable for a Coq workshop project, for example.

It might be smaller than that: as part of the evaluation for the Alectryon paper I has done a port of the reference manual.  It's still up [here](https://alectryon-paper.github.io/bench/refman/).  I'm sure there are imperfections, of course, but the point was that it very much could be done.
